### PR TITLE
Lint and format code in CI and pre-commit

### PR DESCRIPTION
**Pull request by bolinocroustibat**
- Add modern [Ruff](https://docs.astral.sh/ruff/) linter and formatter with default config including [default Flake8's linting rules](https://docs.astral.sh/ruff/rules/) and [default Black formatting rules](https://black.readthedocs.io/en/stable/the_black_code_style/current_style.html). This replaces flake8 (linter), Black (formatter), pylint and isort.
- [Add lint and format step in CI](https://github.com/opendatateam/udata/pull/3085/files#diff-78a8a19706dbd2a4425dd72bdab0502ed7a2cef16365ab7030a5a0588927bf47) EDIT: without fixing
- [Add pre-commit hook to lint and format code](https://github.com/opendatateam/udata/pull/3085/files#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9)
- [Update contributing documentation](https://github.com/opendatateam/udata/pull/3085/files#diff-eca12c0a30e25b4b46522ebf89465a03ba72a03f540796c979137931d8f92055)
- Lint, format and sort imports on udata codebase

**Remark 1:**
Since this PR is heavy, I suggest checking the changes/diff **before** the commit 6d1b9aa (_"style: lint and format code and sort imports"_), this commit being the one dedicated to lint and format the code and making the whole PR barely readable :)

**Remark 2:**
I noticed the team might prefer single quotes (') over double quotes ("). Ruff/Flake8 default rule is to use double quotes, but we can configure a special rule for to use single quotes if the team prefers.
In that case, we can either use a `ruff.toml` to specify this special formatting rule:
```toml
[format]
quote-style = "single"
```
...or a `pyproject.toml` config file if we want to use a general Python package config file (so the config file can be used for many other config, and/or for Python packaging):
```toml
[tool.ruff.format]
quote-style = "single"
```

Let me know if you want to add those special styles and/or any other, I can modify the code format accordingly in this same PR.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -38,6 +38,12 @@ jobs:
           key: py3-cache-v12-{{ arch }}-{{ .Branch }}
           paths:
           - venv
+      - run:
+          name: Lint and format code and sort imports
+          command: |
+            source venv/bin/activate
+            ruff check --fix --select I .
+            ruff format .
```
Comment (maudetes) [inline .circleci/config.yml:46]: I'm not sure I see the results of each command separately.

**Review by maudetes — COMMENTED** [review]
Thank you! I think this is much needed :) Ruff seems indeed quite efficient.

Some general comments :
- flake8 config in setup.cfg should be removed?
- I think keeping single quotes would allow a lesser impact on git history
    - I would go with the pyproject approach for the config
- on a second iteration, we can add this config on all udata plugins I think :rocket:

File: .pre-commit-config.yaml [hunk 2]
```diff
@@ -22,3 +23,23 @@ repos:
...
+    hooks:
+      - id: check-yaml
+      - id: end-of-file-fixer
+      - id: trailing-whitespace
+      - id: check-added-large-files
+
+  - repo: https://github.com/astral-sh/ruff-pre-commit
+    rev: v0.5.1 # Ruff version
+    hooks:
+      - id: ruff
```
[the remark this answers is not in the document]
Comment (bolinocroustibat) [reply .pre-commit-config.yaml:39]: Indeed, thanks for spotting. Added to the PR.

File: .circleci/config.yml (hunk 1 above)
[the remark this answers is not in the document]
Comment (bolinocroustibat) [reply .circleci/config.yml]: Indeed, done.

**Pull request merged**